tools/github/release/manager.py

```diff
@@ -0,0 +1,143 @@
+        self._log = log
+        self.oauth_token = oauth_token
+        self.user = user
+        self._asset_types = asset_types
+        self._github = github
+        self._session = session
+        self.create = create
+
+    async def __aenter__(self) -> AGithubReleaseManager:
+        return self
+
+    async def __aexit__(self, *args) -> None:
+        await self.close()
+
+    def __getitem__(self, version) -> AGithubRelease:
+        # return self.release_class(self, version)
+        raise NotImplementedError
+
+    @property
+    def release_class(self) -> Type[AGithubRelease]:
+        # return GithubRelease
+        raise NotImplementedError
+
+    @cached_property
+    def github(self) -> gidgethub.abc.GitHubAPI:
+        return (
+            self._github
+            or gidgethub.aiohttp.GitHubAPI(self.session, self.user, oauth_token=self.oauth_token))
+
+    @cached_property
+    def log(self) -> verboselogs.VerboseLogger:
+        return self._log or verboselogs.VerboseLogger(__name__)
+
+    @cached_property
+    def path(self) -> pathlib.Path:
+        return pathlib.Path(self._path)
+
+    @async_property
+    async def latest(self) -> Dict[str, packaging.version.Version]:
+        latest = {}
+        for release in await self.releases:
+            version = self.parse_version(release["tag_name"])
+            if not version:
+                continue
+            latest[str(version)] = version
+            minor = f"{version.major}.{version.minor}"
+            if version > latest.get(minor, self.version_min):
+                latest[minor] = version
+        return latest
+
+    @async_property
+    async def releases(self) -> List[Dict]:
+        results = []
+        async for result in self.github.getiter(str(self.releases_url)):
+            results.append(result)
+        return results
+
+    @cached_property
+    def releases_url(self) -> pathlib.PurePosixPath:
+        return pathlib.PurePosixPath(f"/repos/{self.repository}/releases")
+
+    @cached_property
+    def session(self) -> aiohttp.ClientSession:
+        return self._session or aiohttp.ClientSession()
+
+    @property
+    def version_min(self) -> packaging.version.Version:
+        return VERSION_MIN
+
+    @cached_property
+    def version_re(self) -> Pattern[str]:
+        return re.compile(self._version_re)
+
+    async def close(self) -> None:
+        if not "session" in self.__dict__:
+            return
+        await self.session.close()
+        del self.__dict__["session"]
+
+    def fail(self, message: str) -> str:
+        if not self.continues:
+            raise GithubReleaseError(message)
+        self.log.warning(message)
+        return message
+
+    def format_version(self, version: Union[str, packaging.version.Version]) -> str:
+        return self._version_format.format(version=version)
+
+    def parse_version(self, version: str) -> Optional[packaging.version.Version]:
+        _version = self.version_re.sub(r"\1", version)
```

Comment:
I really wish we had a linter for the `_` locals..

Reply:
im thankful we dont - but happy to clean these up...